### PR TITLE
refactor: decouple executor from storage layer

### PR DESCRIPTION
**Pull request by retran**
## Why

After PRs #235–#237, the executor still holds direct references to `*mpr.Reader` and `*mpr.Writer` in its struct, and several command files still construct BSON for datagrids, filters, and widget cloning. This PR completes the decoupling: the executor works primarily through the `backend.FullBackend` abstraction.

This is the architectural pivot — dependency inversion via a `BackendFactory` pattern replaces the direct MPR reader/writer construction.

## What changed (incremental from PR #237)

**32 files changed, +1,523/-3,121 (net -1,598 lines)**

**Executor struct decoupled:**
- Removed `writer` and `reader` fields from the `Executor` struct
- Added `BackendFactory` injection — `executor_connect.go` rewritten to create backends through the factory instead of constructing `mpr.Reader`/`mpr.Writer` directly
- `cmd/mxcli/main.go` updated to inject the factory

**Remaining BSON extracted:**
- `mdl/backend/mpr/datagrid_builder.go` (1,260 lines) — DataGrid2 construction, filter widgets, column building moved from executor to backend
- `mdl/backend/mpr/datagrid_builder_test.go` — test coverage for datagrid building

**Deleted from executor (5 files, ~1,800 lines):**
- `bson_helpers.go` — BSON utility functions
- `cmd_pages_builder_input_cloning.go` — widget cloning logic
- `cmd_pages_builder_input_datagrid.go` — datagrid construction
- `cmd_pages_builder_input_filters.go` — filter widget construction
- `cmd_pages_builder_v3_pluggable.go` — pluggable widget BSON

**Result:** The only BSON remaining in the executor is in 3 read-only files (describe, diff) that deserialize for display — no write-path BSON. `Reader()` remains as an escape hatch for these read-only paths.

## Review-driven improvements

- **Mock defaults:** `OpenPageForMutation`, `OpenWorkflowForMutation`, `BuildDataGrid2Widget`, `BuildFilterWidget` now return descriptive errors instead of `(nil, nil)`
- **Nil guards:** `updateWidgetsInContainer` checks for nil mutator; `ParseMicroflowFromRaw` returns stub MDL on nil
- **Spec type aliasing:** `NavigationProfileSpec`, `NavHomePageSpec`, `NavMenuItemSpec` in `sdk/mpr` are now type aliases of `mdl/types` — eliminates silent divergence risk
- **Field-count drift assertions:** `TestFieldCountDrift` with 30 assertions (15 struct pairs) catches field additions/removals in manually-copied types
- **Disconnect error propagation:** `execConnect`, `reconnect`, `execDisconnect` log disconnect errors; `Executor.Close()` returns disconnect error
- **Nil backendFactory guard:** `reconnect` checks for nil factory before attempting reconnection
- **Stale comments:** "underlying reader" → "backend" in `cmd_pages_builder.go`
- **Unused param:** `buildDataGrid2Property` receiver renamed to `_`

## Stack

PR 4/5 in the shared-types refactoring stack.
Merge order: #232 → #235 → #236 → #237 → **this** → #239

### CODE REVIEW

File: mdl/executor/cmd_widgets.go [hunk 1]
```diff
@@ -197,99 +195,35 @@ func groupWidgetsByContainer(widgets []widgetRef) map[string][]widgetRef {
...
+	// Open the container (page, layout, or snippet) through the backend mutator.
+	mutator, err := ctx.Backend.OpenPageForMutation(model.ID(containerID))
 	if err != nil {
-		return 0, mdlerrors.NewBackend(fmt.Sprintf("load snippet %s", containerName), err)
-	}
-	var rawData bson.D
-	if err := bson.Unmarshal(rawBytes, &rawData); err != nil {
-		return 0, mdlerrors.NewBackend(fmt.Sprintf("unmarshal snippet %s", containerName), err)
+		return 0, mdlerrors.NewBackend(fmt.Sprintf("open %s for mutation", containerName), err)
 	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/cmd_widgets.go:211]: Fixed in 66f637a — added nil guard that returns `mdlerrors.NewBackend` with descriptive message.

File: mdl/backend/mock/mock_mutation.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+// ---------------------------------------------------------------------------
+// PageMutationBackend
+// ---------------------------------------------------------------------------
+
+func (m *MockBackend) OpenPageForMutation(unitID model.ID) (backend.PageMutator, error) {
+	if m.OpenPageForMutationFunc != nil {
+		return m.OpenPageForMutationFunc(unitID)
+	}
+	return nil, nil
+}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/backend/mock/mock_mutation.go:23]: Fixed in 66f637a — mock now returns `fmt.Errorf("MockBackend.<Method> not configured")` instead of `(nil, nil)`.

File: mdl/bsonutil/bsonutil.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+func IDToBsonBinary(id string) primitive.Binary {
+	blob := types.UUIDToBlob(id)
+	if blob == nil || len(blob) != 16 {
+		blob = types.UUIDToBlob(types.GenerateID())
+	}
+	return primitive.Binary{
+		Subtype: 0x00,
+		Data:    blob,
+	}
+}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/bsonutil/bsonutil.go:24]: This is intentional fail-fast behavior — IDToBsonBinary panics on invalid UUID. An error-returning `IDToBsonBinaryErr` variant is tracked as a deferred item for a separate PR.

File: mdl/executor/cmd_pages_builder.go [hunk 4]
```diff
@@ -61,17 +62,25 @@ func (pb *pageBuilder) initPluggableEngine() {
...
 // registerWidgetName registers a widget name and returns an error if it's already used.
 // Widget names must be unique within a page/snippet.
+
+// getProjectPath returns the project directory path from the underlying reader.
+func (pb *pageBuilder) getProjectPath() string {
+	if pb.backend != nil {
+		return pb.backend.Path()
+	}
+	return ""
+}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/cmd_pages_builder.go]: Fixed in 66f637a — comment now says "backend".

File: mdl/backend/mpr/datagrid_builder.go [hunk 5]
```diff
@@ -0,0 +1,1260 @@
...
+		}
+	}
+	return result
+}
+
+// ===========================================================================
+// BSON property builders (package-level, no receiver needed)
+// ===========================================================================
+
+func buildDataGrid2Property(entry pages.PropertyTypeIDEntry, datasource pages.DataSource, attrRef string, primitiveValue string, b *MprBackend) bson.D {
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/backend/mpr/datagrid_builder.go]: Fixed in 3d79a42 — renamed to `_` since the parameter is unused.

File: mdl/backend/mock/mock_mutation.go [hunk 6]
```diff
@@ -94,3 +96,17 @@ func (m *MockBackend) BuildCreateAttributeObject(attributePath string, objectTyp
...
+		return m.BuildDataGrid2WidgetFunc(id, name, spec, projectPath)
+	}
+	return nil, nil
+}
+
+func (m *MockBackend) BuildFilterWidget(spec backend.FilterWidgetSpec, projectPath string) (pages.Widget, error) {
+	if m.BuildFilterWidgetFunc != nil {
+		return m.BuildFilterWidgetFunc(spec, projectPath)
+	}
+	return nil, nil
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/backend/mock/mock_mutation.go]: Fixed in 3d79a42 — both now return `fmt.Errorf("MockBackend.<Method> not configured")`, consistent with the other mock methods.

File: mdl/executor/executor_connect.go [hunk 7]
```diff
@@ -63,20 +60,18 @@ func reconnect(ctx *ExecContext) error {
...
-		e.writer.Close()
+	if e.backend != nil && e.backend.IsConnected() {
+		e.backend.Disconnect()
 	}
 
 	// Reopen connection
-	writer, err := mpr.NewWriter(e.mprPath)
-	if err != nil {
+	b := e.backendFactory()
+	if err := b.Connect(e.mprPath); err != nil {
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/executor_connect.go:76]: Fixed in 3d79a42 — added nil guard matching `execConnect`.

File: mdl/executor/executor.go [hunk 8]
```diff
@@ -295,21 +302,29 @@ func (e *Executor) Catalog() *catalog.Catalog {
...
+	if e.backend == nil {
+		return nil
+	}
+	type readerProvider interface {
+		MprReader() *mpr.Reader
+	}
+	if rp, ok := e.backend.(readerProvider); ok {
+		return rp.MprReader()
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/executor.go:316]: Good catch — updated PR description. `Reader()` is an intentional transitional escape hatch for linter rules and edge cases that haven't been migrated to the backend interface yet. The comment on the method documents this.

**Comment by retran** [comment]
Thanks for the review.

1. **Microflow null check** — agreed, the hardcoded string format is intentional here. This is a describe/diff context where we render a stub MDL representation when parsing fails. Propagating an error would suppress the rest of the diff output.

2. **MprReader() escape hatch** — correct, this is a deliberate transitional API. The comment on the method documents the intent and advises callers to prefer Backend methods. It will be removed once linter rules and remaining edge cases are migrated.

File: mdl/executor/executor_connect.go [hunk 9]
```diff
@@ -6,30 +6,27 @@ import (
...
-	"github.com/mendixlabs/mxcli/sdk/mpr"
 )
 
 func execConnect(ctx *ExecContext, s *ast.ConnectStmt) error {
 	e := ctx.executor
-	if ctx.ConnectedForWrite() {
-		e.writer.Close()
+	if e.backend != nil && e.backend.IsConnected() {
+		e.backend.Disconnect()
 	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/executor_connect.go:18]: Fixed in de25c84 — now logs warning via `fmt.Fprintf(ctx.Output, ...)`.

File: mdl/executor/executor_connect.go [hunk 10]
```diff
@@ -63,20 +60,21 @@ func reconnect(ctx *ExecContext) error {
...
 
 	// Close existing connection
-	if ctx.ConnectedForWrite() {
-		e.writer.Close()
+	if e.backend != nil && e.backend.IsConnected() {
+		e.backend.Disconnect()
 	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/executor_connect.go:69]: Fixed in de25c84 — same pattern, logs warning before proceeding with reconnect.

File: mdl/executor/executor_connect.go [hunk 11]
```diff
@@ -103,10 +101,8 @@ func execDisconnect(ctx *ExecContext) error {
 		fmt.Fprintf(ctx.Output, "Warning: finalization error: %v\n", err)
 	}
 
-	e.writer.Close()
+	e.backend.Disconnect()
 	fmt.Fprintf(ctx.Output, "Disconnected from: %s\n", e.mprPath)
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/executor_connect.go]: Fixed in de25c84 — logs warning matching the `finalizeProgramExecution` pattern above it.

File: mdl/executor/executor.go [hunk 12]
```diff
@@ -295,21 +302,29 @@ func (e *Executor) Catalog() *catalog.Catalog {
...
 // Close closes the connection to the project and all SQL connections.
 func (e *Executor) Close() error {
-	if e.writer != nil {
-		e.writer.Close()
-		e.writer = nil
-		e.reader = nil
+	if e.backend != nil && e.backend.IsConnected() {
+		e.backend.Disconnect()
 		e.backend = nil
 	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/executor.go:330]: Fixed in 4d2582a — `Close()` now returns the error from `Disconnect()`.

**Comment by ako** [comment]
## Code Review

### Overview

This PR completes the executor decoupling that PRs #235–#237 began. Net result: the executor holds no `*mpr.Reader` or `*mpr.Writer` fields, five BSON-heavy executor files (~1,800 lines) are deleted, and DataGrid2/filter/cloning logic moves to `mdl/backend/mpr/datagrid_builder.go`. It also addresses two specific issues raised in the PR #235 review: the `NavigationProfileSpec` divergence risk and the absence of field-drift detection.

---

### Positive Aspects

- **BackendFactory pattern is clean.** `executor_connect.go` replaces the old triple-assignment (`writer`/`reader`/`backend = mprbackend.Wrap(...)`) with a straightforward factory call + `Connect()`. Nil guard on `backendFactory` is present. Wiring in `cmd/mxcli/main.go` is correct.
- **Spec type aliases close the divergence risk flagged in #235.** `NavigationProfileSpec`, `NavHomePageSpec`, `NavMenuItemSpec` in `sdk/mpr` are now proper type aliases (`= types.*`). The `unconvertNavProfileSpec` conversion collapses to a pass-through, removing 27 lines of manual field copying.
- **`TestFieldCountDrift` is a good lightweight guard.** 30 assertions across 15 struct pairs using `reflect.TypeOf(v).NumField()` will catch any field addition/removal in manually-converted types and force an explicit update to `convert.go`.
- **Mock defaults are now fail-loud.** `OpenPageForMutation`, `OpenWorkflowForMutation`, `BuildDataGrid2Widget`, `BuildFilterWidget` all return descriptive `"MockBackend.X not configured"` errors instead of `(nil, nil)` — prevents silent nil-dereference bugs in tests.
- **`Reader()` escape hatch is handled well.** Uses a local `readerProvider` interface to avoid circular imports, has nil guards, and carries a deprecation comment signalling intent to remove it.

---

### Issues & Concerns

**1. `datagrid_builder_test.go` lacks happy-path coverage for the main entry points**

The test file covers only the deep-clone helper (three tests: ID regeneration, arrays, nil preservation). There are no tests for:
- `BuildDataGrid2Widget()` — column assembly, datasource binding, paging, selection
- `BuildFilterWidget()` — any filter type
- The property builders (`buildDataGrid2Property`, `buildFiltersPlaceholderProperty`, `applyDataGridPagingProps`, `applyDataGridSelectionProp`)

These are the highest-value paths in the new file and the most likely to have edge cases. The PR moves them from the executor (where they were also untested) to the backend — a good structural move — but doesn't add new coverage in the process.

**2. `TestFieldCountDrift` validates cardinality only, not field names or order**

If a field is renamed (e.g., `Name` → `DisplayName`) the count stays the same and the test passes while `convert.go` silently copies the wrong value. This is inherent to the approach but worth noting — it's a partial guard, not a complete one.

**3. New `WidgetBuilderBackend` interface methods are not in the mock by default**

`BuildDataGrid2Widget` and `BuildFilterWidget` are added to `mock_mutation.go` with descriptive error defaults (good), but there are no `MockDataGrid2Mutator` helpers or test fixtures for exercising these methods. Tests that exercise datagrid creation paths in the executor still need integration with a real backend.

---

### Checklist Against Project Standards

| Item | Status |
|---|---|
| Executor no longer imports `sdk/mpr` for write paths | ✅ Only 3 read-only display files remain |
| BackendFactory nil guard | ✅ Present in `reconnect` |
| Spec type divergence resolved | ✅ Aliases in `sdk/mpr/writer_navigation.go` |
| Field drift detection | ✅ `TestFieldCountDrift` with 30 assertions |
| Mock defaults fail-loud | ✅ Descriptive error messages |
| `datagrid_builder.go` happy-path tests | ⚠️ Only cloning helper tested; widget construction untested |
| Error propagation improvements | ✅ `Close()` returns error, disconnect errors logged |

---

### Verdict

**Approve.** The BackendFactory pattern is well-implemented, the spec aliasing fix directly addresses prior review feedback, and the field-drift test is a practical addition. The main gap is that `datagrid_builder.go`'s primary entry points (`BuildDataGrid2Widget`, `BuildFilterWidget`) have no unit tests — these moved from executor to backend untested. That's worth a follow-up issue rather than a blocker, since the structural goal of this PR is achieved cleanly.
